src/shared/dissect-image.c

```diff
@@ -1727,39 +1856,76 @@ int verity_settings_load(
 
         assert(verity);
         assert(image);
+        assert(verity->designator < 0 || IN_SET(verity->designator, PARTITION_ROOT, PARTITION_USR));
 
         /* If we are asked to load the root hash for a device node, exit early */
         if (is_device_path(image))
                 return 0;
 
+        designator = verity->designator;
+
         /* We only fill in what isn't already filled in */
 
         if (!verity->root_hash) {
                 _cleanup_free_ char *text = NULL;
 
                 if (root_hash_path) {
+                        /* If explicitly specified it takes precedence */
                         r = read_one_line_file(root_hash_path, &text);
                         if (r < 0)
                                 return r;
+
+                        if (designator < 0)
+                                designator = PARTITION_ROOT;
                 } else {
-                        r = getxattr_malloc(image, "user.verity.roothash", &text, true);
-                        if (r < 0) {
-                                _cleanup_free_ char *p = NULL;
+                        /* Otherwise look for xattr and separate file, and first for the data for root and if
+                         * that doesn't exist for /usr */
 
-                                if (!IN_SET(r, -ENODATA, -ENOENT) && !ERRNO_IS_NOT_SUPPORTED(r))
-                                        return r;
+                        if (designator < 0 || designator == PARTITION_ROOT) {
+                                r = getxattr_malloc(image, "user.verity.roothash", &text, true);
+                                if (r < 0) {
+                                        _cleanup_free_ char *p = NULL;
 
-                                p = build_auxiliary_path(image, ".roothash");
-                                if (!p)
-                                        return -ENOMEM;
+                                        if (!IN_SET(r, -ENODATA, -ENOENT) && !ERRNO_IS_NOT_SUPPORTED(r))
+                                                return r;
 
-                                r = read_one_line_file(p, &text);
-                                if (r < 0 && r != -ENOENT)
-                                        return r;
+                                        p = build_auxiliary_path(image, ".roothash");
+                                        if (!p)
+                                                return -ENOMEM;
+
+                                        r = read_one_line_file(p, &text);
+                                        if (r < 0 && r != -ENOENT)
+                                                return r;
+                                }
+
+                                if (text)
+                                        designator = PARTITION_ROOT;
+                        }
+
+                        if (!text && (designator < 0 || designator == PARTITION_USR)) {
+                                r = getxattr_malloc(image, "user.verity.usrhash", &text, true);
```

Comment:
Interesting naming choice - in dm-verity literature, the "root" in root hash usually refers to the root of the merkle tree, as you probably know already - but here it's changing into a reference to the mount point. Which is perfectly fine - maybe worth a mention in the documentation?

Reply:
I'll add a comment about the lucky naming coincidence to the sources, but I don't think we should document this beyond that in the user-facing docs, people who know the background will probably either laugh inside, or die inside, and the others we shouldn't confuse I guess...

Reply:
Eheh, fair enough